pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/KafkaProducer.java

```diff
@@ -0,0 +1,364 @@
+          .setBatchingSettings(BatchingSettings.newBuilder()
+              .setElementCountThreshold(elementCount)
+              .setRequestByteThreshold((long) batchSize)
+              .setDelayThreshold(Duration.ofMillis(lingerMs))
+              .build())
+          .setRetrySettings(RetrySettings.newBuilder()
+              .setMaxAttempts(retries)
+              .setRetryDelayMultiplier(MULTIPLIER)
+              .setInitialRetryDelay(Duration.ofMillis(retriesMs))
+              .setMaxRetryDelay(Duration.ofMillis((retries + 1) * retriesMs))
+              .setRpcTimeoutMultiplier(MULTIPLIER)
+              .setInitialRpcTimeout(Duration.ofSeconds(TOTAL_DELAY))
+              .setMaxRpcTimeout(Duration.ofSeconds(TOTAL_DELAY))
+              .setTotalTimeout(Duration.ofSeconds(TOTAL_DELAY))
+              .build())
+          .build();
+    } catch (IOException e) {
+      throw new KafkaException("Failed to construct kafka producer", e);
+    }
+    return pub;
+  }
+
+  /**
+   * Sends the given record.
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+    return send(record, null);
+  }
+
+  //TODO: there maybe a race condition/starvation between flush and send.
+  //TODO: mimic all kafka's exceptions.
+
+  /**
+   * Sends the given record and invokes the specified callback.
+   * The given record must have the same topic as the producer.
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+    if (closed.get())
+      throw new IllegalStateException("Cannot send after the producer is closed.");
+
+    if (record == null)
+      throw new NullPointerException();
+
+    final Callback cb = callback;
+
+    byte[] valueBytes = ByteString.EMPTY.toByteArray();
+    if (record.value() != null) {
+      try {
+        valueBytes = valueSerializer.serialize(record.topic(), record.value());
+      } catch (ClassCastException e) {
+        throw new SerializationException("Can't convert value of class " +
+            record.value().getClass().getName() + " to class " +
+            producerConfig.getClass(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in value.serializer");
+      }
+    }
+
+    if (valueBytes.length == 0)
+      throw new NullPointerException("Value cannot be null or an empty string.");
+
+    Map<String, String> attributes = new HashMap<>();
+
+    byte[] keyBytes = ByteString.EMPTY.toByteArray();
+    if (record.key() != null) {
+      try {
+        keyBytes = keySerializer.serialize(record.topic(), record.key());
+      } catch (ClassCastException cce) {
+        throw new SerializationException("Can't convert key of class " +
+            record.key().getClass().getName() + " to class " +
+            producerConfig.getClass(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in key.serializer");
+      }
+    }
+    attributes.put("key", new String(keyBytes));
```

Comment:
Yeah, I figured this out while pulling through the console - but didn't know which option to go with -.
What is the appropriate thing to do if the key exceeded this threshold? I tried to find any analogous config in Kafka, but it's all about the message size (payload).